### PR TITLE
Support `jaeger-baggage` header for ad-hoc baggage

### PR DESCRIPTION
**Pull request by yurishkuro**
Signed-off-by: Yuri Shkuro <ys@uber.com>

### CODE REVIEW

File: jaeger_client/codecs.py [hunk 1]
```diff
@@ -110,15 +113,27 @@ def extract(self, carrier):
                 if self.url_encoding:
                     value = urllib_parse.unquote(value)
                 debug_id = value
-        if not trace_id and baggage:
-            raise SpanContextCorruptedException('baggage without trace ctx')
```
Comment (yurishkuro) [inline jaeger_client/codecs.py]: I could not find any rationalization for why this exception had to be thrown in this scenario.

File: tests/test_codecs.py [hunk 2]
```diff
@@ -208,19 +208,6 @@ def test_context_from_readable_headers(self):
                 'hermes': 'LaBarbara Hermes',
             }
 
-    def test_baggage_without_trace_id(self):
```
Comment (yurishkuro) [inline tests/test_codecs.py]: moved down (outside of class) to be able to get `tracer` fixture. Not sure if the same possible when subclassing `TestCodecs(unittest.TestCase)`.

File: tests/test_thrift.py [hunk 3]
```diff
@@ -69,10 +70,6 @@ def serialize(span_id):
         span.finish()
         _marshall_span(span)
 
-    trace_id = 0
```
Comment (yurishkuro) [inline tests/test_thrift.py]: how was this testing "large trace IDs"? 0 wasn't even a valid id.

File: jaeger_client/tracer.py [hunk 4]
```diff
@@ -157,10 +157,12 @@ def start_span(self,
                     tags = tags or {}
                     for k, v in six.iteritems(sampler_tags):
                         tags[k] = v
-            elif self.is_debug_allowed(operation_name):  # have debug id
+            elif parent.debug_id and self.is_debug_allowed(operation_name):
```
Comment (isaachier) [inline jaeger_client/tracer.py]: I think this conflates `debug_id` with debug flag. A debug ID is a specific header along the lines of `jaeger-debug-id: foo`, which then associates an ID `"foo"` with a new span. That span also sets its debug flag to true (assuming throttling allows it). In retrospect, maybe we should revert this change.

File: jaeger_client/tracer.py [hunk 5]
```diff
@@ -157,10 +157,12 @@ def start_span(self,
...
                 flags = SAMPLED_FLAG | DEBUG_FLAG
                 tags = tags or {}
                 tags[self.debug_id_header] = parent.debug_id
+            if parent and parent.baggage:
+                baggage = dict(parent.baggage)  # TODO do we need to clone?
```
Comment (isaachier) [inline jaeger_client/tracer.py]: If the `dict` is immutable, you should not have to clone. Immutability is definitely the approach in other clients, IDK about this one.

File: jaeger_client/tracer.py (hunk 4 above)
Comment (yurishkuro) [reply jaeger_client/tracer.py]: where do you se a disconnect? The only way we get to this part is if we have no parent OR the parent is empty but has a debug_id tag set. Previously those were the only condition, now you can also get here with the parent empty (as in: no trace id), but having the baggage. Yet it may not have the debug_id tag. Previously this `elif` part always implied there are the debug_id tag, now we need to check for it.

Comment (isaachier) [reply jaeger_client/tracer.py]: OK sorry my confusion here was based on a misunderstanding here regarding the necessity of a parent. I had thought you were implying that all debug flags must be propagated from a real span object, not just a header. I forgot the header is transformed into an empty span context with a debug ID, but nothing else.
